### PR TITLE
CC: Manual Work is a Bug (#1805)

### LINKED ISSUE #1805: CC: Manual Work is a Bug
**Author:** @bartlettroscoe 
**EB Member:** @pagrubel 

From David Rogers (@frobnitzem)  ...

I just stumbled across this article that lays out a good explanation of the reproducibility, documentation, and automation mindset:
"Manual Work is a Bug"

https://queue.acm.org/detail.cfm?id=3197520

The same author has a few more good (and also some hilarious) articles in the automation/devops theme: 

https://www.researchgate.net/scientific-contributions/Thomas-A-Limoncelli-2139171841

### PR DESCRIPTION
**Pull request by bartlettroscoe**
# Description

EB Member: @pagrubel 

Resolves #1805

## PR checklist for files displayed on bssw.io site

* [x] ***[Author]*** `@mention` the BSSw.io editorial board member `@<eb-member-id>` in **Description** above assigned to shepherd your PR.
* [x] ***[Author]*** Add the `<issue-id>` in the **Description** above for the associated GitHub Issue.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py])~~.
* [x] ***[Author]*** Inspect the content in the `*.md` file(s) as rendered in GitHub for this PR.
* [x] ***[EB Mem]*** Assign this PR to the EB member `<eb-member-id>`.
* [x] ***[EB Mem]*** Assign this PR to the author of the PR `<pr-author-id>`.
* [x] ***[EB Mem]*** Add one or more Reviewers.
* [x] ***[EB Mem]*** Add label `content: <content-type>` for the type of contribution.
* [x] ***[EB Mem]*** Add to Project `Content Development` (see [Content Development]).
* [x] ***[EB Mem]*** Add [meta-data] to the `*.md` file(s) (set `Publish: yes`).
* [x] ***[EB Mem]*** Add label `preview` (so PR branch will be merged to 'preview' branch and watch for possible merge failures).
* [x] ***[EB Mem]*** Rebuild [preview] site and confirm new content is there, renders correctly and is returned in searches.
* [x] ***[Author]*** Make any final changes to the PR based on feedback.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).~~
* [x] ***[EB Mem]*** Rebuild [preview] site and re-confirm content looks correct.
* [x] ***[EB Mem]*** Ensure at least one reviewer signs off on the final changes.
* [x] ***[EB Mem]*** Move the PR to "Ready to Publish" in [Content Development].
* [x] ***[EB Mem]*** Leave comment and @mention `@betterscientificsoftware/bssw-maint` (**BSSw Maint**) asking to carry out final publication steps.
* [x] ***[BSSw Maint]*** Verify that all needed files are present in the PR (article, images, updates to Site/Homepage.md carousel and/or Site/Announcements/Announcements.md as appropriate).
* [x] ***[BSSw Maint]*** Merge PR. (Should automatically move to "Done" in [Content Development].)
* [x] ***[BSSw Maint]*** Verify new contribution shows up on [bssw.io] as expected.

NOTE:
* Checklist items prefixed with ***[Author]*** are expected to be performed by the author of the PR or can be performed by the author.
* Checklist items prefixed with ***[EB Mem]*** must be performed by a [BSSw.io Editorial Board (EB) Member](https://betterscientificsoftware.github.io/bssw.io/bssw_members.html).
* Checklist items prefixed with ***[BSSw Maint]*** must be performed by someone in the `@betterscientificsoftware/bssw-maint` team (hint: type `@`, `b`, `s`, `s`, `w`, `-`, `m`  to auto-complete to `@betterscientificsoftware/bssw-maint`).

[preview]: https://preview.bssw.io
[bssw.io]: https://bssw.io
[Content Development]: https://github.com/betterscientificsoftware/bssw.io/projects/3
[BSSw Internal]: https://github.com/betterscientificsoftware/bssw.io/projects/2
[meta-data]: https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#metadata-section
[wikize_refs.py]: https://github.com/betterscientificsoftware/bssw.io/blob/master/utils/README.md#wikize_refspy

### CODE REVIEW

**Comment by pagrubel** [comment]
@bartlettroscoe I have a few corrections I'll post in a bit.

File: CuratedContent/ManualWorkIsABug.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+:--- | :---
+Paper Title | Manual Work is a Bug
+Authors | Thomas A. Limoncelli
+Publication | [ACM Queue, March 14, 2018, Volume 16, Issue 1](https://queue.acm.org/detail.cfm?id=3197520)
+
+Nearly everyone who works with computers often has to perform routine tasks.
+Some people like system administrators (sysadmins) and DevOps engineers have **many** such tasks to perform.
+However, even other types of developers and users often have repetitive tasks that they need to execute repeatedly.
+And others in their teams and organizations often need to perform similar processes/tasks (or the same tasks, with minor variations).
+These routine processes may take long enough, be complex enough, have severe enough consequences if they are done in correctly, and/or be likely to be repeated enough to justify automating them to some degree.
```
Comment (pagrubel) [inline CuratedContent/ManualWorkIsABug.md:22]:
```suggestion
These routine processes may be time-consuming, be complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.
```

File: CuratedContent/ManualWorkIsABug.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+Thomas A. Limoncelli's article "Manual Work is a Bug" describes an Agile incremental process for developing increasing levels of automation for routine processes like these.
+(Thomas mainly speaks to sysadmins in this article, but the process and the principles described are the same for any individual or team that needs to perform routine tasks with computers.)
+
+The basic idea is to start by writing documentation right from the beginning while manually performing the steps for the first time for every process that is a potential candidate to become a routine (automatable) process.
```
Comment (pagrubel) [inline CuratedContent/ManualWorkIsABug.md:26]:
```suggestion
The basic idea is to start by writing documentation right from the beginning while manually performing the steps, for the first time, for every process that is a potential candidate to become a routine (automated) process.
```

File: CuratedContent/ManualWorkIsABug.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+* **Phase 1: Document the steps**: Document the process as you do the steps manually the first time.
+
+* **Phase 2: Create automation equivalents**: When executing the process again, go back over the process manually following the existing documentation while refining the process and adding exact command-line snippets to the documentation.
+
+* **Phase 3: Create automation**: Develop an expanding set of scripts/code to automate parts of the process as it is executed repeatedly.
+(Every process iteration should expand the automation scripts/code scope and reduce the manual effort.)
+
+* **Phase 4: Self-service autonomous systems**:  The final phase is to turn the process into a stand-alone tool using the developed scripts/code that runs with a single invocation.
+In some cases, such processes justify being set up as an ***autonomous process** (i.e., a process that runs automatically without any explicit action by the user.)
```
Comment (pagrubel) [inline CuratedContent/ManualWorkIsABug.md:44]:
```suggestion
In some cases, such processes justify being set up as an **autonomous process** (i.e., a process that runs automatically without any explicit action by the user.)
```

**Review by pagrubel — CHANGES_REQUESTED** [review]
@bartlettroscoe  Looks good I've just requested some minor changes. Use what you want.

File: CuratedContent/ManualWorkIsABug.md (hunk 1 above)
Comment (bartlettroscoe) [reply CuratedContent/ManualWorkIsABug.md:22]: I was trying to emphasize that there my be thresholds for time, complexity, severity, and/or future repeatability to trigger considering automating the process.   Does the updated text make that clear?

Comment (bartlettroscoe) [reply CuratedContent/ManualWorkIsABug.md]: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa (except I removed "be" from "be complex").

File: CuratedContent/ManualWorkIsABug.md (hunk 2 above)
Comment (bartlettroscoe) [reply CuratedContent/ManualWorkIsABug.md]: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa.

File: CuratedContent/ManualWorkIsABug.md (hunk 3 above)
Comment (bartlettroscoe) [reply CuratedContent/ManualWorkIsABug.md]: Applied in commit 34a11ccbd84db1ea44bcf3eaea9c31c584cfe4fa.

**Comment by bartlettroscoe** [comment]
Hello @pagrubel, I committed all of your great suggestions.  Can you please approve the review?  (But I guess that is optional in our workflow).

@rinkug, I believe this CC article is ready to publish.

**Comment by bartlettroscoe** [comment]
> Hello @pagrubel, I committed all of your great suggestions. Can you please approve this PR? (But I guess that is optional in our workflow.)

Hello @pagrubel, actually, the PR is block from merge until the request for changes review is updated or dismissed.

File: CuratedContent/ManualWorkIsABug.md [hunk 4]
```diff
@@ -19,11 +19,11 @@ Nearly everyone who works with computers often has to perform routine tasks.
 Some people like system administrators (sysadmins) and DevOps engineers have **many** such tasks to perform.
 However, even other types of developers and users often have repetitive tasks that they need to execute repeatedly.
 And others in their teams and organizations often need to perform similar processes/tasks (or the same tasks, with minor variations).
-These routine processes may take long enough, be complex enough, have severe enough consequences if they are done in correctly, and/or be likely to be repeated enough to justify automating them to some degree.
+These routine processes may be time-consuming, complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.
```
Comment (pagrubel) [inline CuratedContent/ManualWorkIsABug.md:22]:
```suggestion
These routine processes may be time-consuming or complex, have severe consequences if they are done incorrectly, and/or be likely to be repeated enough to justify automating them to some degree.
```

**Review by pagrubel — COMMENTED** [review]
@bartlettroscoe  Taking out the be didn't completely work so I've suggested another alternative.

**Review by pagrubel — APPROVED** [review]

**Review by pagrubel — APPROVED** [review]

**Comment by pagrubel** [comment]
@bartlettroscoe This is ready to publish. You may want to check the preview site.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe This is ready to publish. You may want to check the preview site.

@pagrubel, already did:

* https://preview.bssw.io/items/manual-work-is-a-bug

😊

(See above checklist items for this.)

**Pull request merged**